packages/xdl/src/start/ExpoUpdatesManifestHandler.ts

```diff
@@ -0,0 +1,148 @@
+import { ExpoUpdatesManifest, getConfig } from '@expo/config';
+import { getRuntimeVersionForSDKVersion } from '@expo/sdk-runtime-versions';
+import express from 'express';
+import http from 'http';
+import { parse } from 'url';
+import uuid from 'uuid';
+
+import {
+  Analytics,
+  Config,
+  ProjectAssets,
+  ProjectUtils,
+  resolveEntryPoint,
+  UrlUtils,
+} from '../internal';
+import {
+  getBundleUrlAsync,
+  getExpoGoConfig,
+  getPackagerOptionsAsync,
+  stripPort,
+} from './ManifestHandler';
+
+function getPlatformFromRequest(req: express.Request | http.IncomingMessage): string {
+  const url = req.url ? parse(req.url, true) : null;
+  const platform = url?.query.platform || req.headers['expo-platform'];
+  if (!platform) {
+    throw new Error('Must specify expo-platform header or query parameter');
+  }
+  return String(platform);
+}
+
+export async function getManifestResponseAsync({
+  projectRoot,
+  platform,
+  host,
+}: {
+  projectRoot: string;
+  platform: string;
+  host?: string;
+}): Promise<{
+  body: ExpoUpdatesManifest;
+  headers: Map<string, number | string | readonly string[]>;
+}> {
+  const headers = new Map<string, any>();
+  // set required headers for expo-updates manifest specification
+  headers.set('expo-protocol-version', 0);
+  headers.set('expo-sfv-version', 0);
+  headers.set('cache-control', 'private, max-age=0');
+  headers.set('content-type', 'application/json');
+
+  const hostname = stripPort(host);
+  const [projectSettings, bundleUrlPackagerOpts] = await getPackagerOptionsAsync(projectRoot);
+  const projectConfig = getConfig(projectRoot);
+  const entryPoint = resolveEntryPoint(projectRoot, platform, projectConfig);
+  const mainModuleName = UrlUtils.stripJSExtension(entryPoint);
+  const expoConfig = projectConfig.exp;
+  const expoGoConfig = await getExpoGoConfig({
+    projectRoot,
+    projectSettings,
+    mainModuleName,
+    hostname,
+  });
+
+  const runtimeVersion =
+    expoConfig.runtimeVersion ??
+    (expoConfig.sdkVersion ? getRuntimeVersionForSDKVersion(expoConfig.sdkVersion) : null);
+  if (!runtimeVersion) {
+    throw new Error('Must specify runtimeVersion or sdkVersion in app.json');
+  }
+
+  const bundleUrl = await getBundleUrlAsync({
+    projectRoot,
+    platform,
+    projectSettings,
+    bundleUrlPackagerOpts,
+    mainModuleName,
+    hostname,
+  });
+
+  // resolve all assets and set them on the manifest as { rawUrl?: string, assetKey?: string } objects
+  const assets = await ProjectAssets.resolveAndCollectExpoUpdatesManifestAssets(
```

Comment:
it might be helpful to add a comment somewhere clarifying that "assets" here refers to "manifest assets" (e.g. `notification.icon`. `splash.image`, etc) rather than assets `require`d in JS. (at least, that's my understanding of what's happening here).

up to now, only the latter have been added to the `assets` field in the new manifest. but i think i understand the reasoning behind this, so that these "manifest assets" all get downloaded onto the device ahead of time by expo-updates, correct?

Reply:
This is doing two things:
- Collect each local asset from `notification.icon`, `splash.image`, etc and place in manifest assets. These are the same assets that are collected during an EAS branch publish.
- Put a new sibling field for each field containing a URL and reference to the asset. For example, for `notification.icon = /path/to/local/image.png` this will add a new field: `notification.iconAsset = { assetKey: <key in assets>, rawUrl: <URL to fetch asset from> }`.

I'll add a comment.